pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentIndexCreationDriverImpl.java

```diff
@@ -105,7 +106,45 @@
   @Override
   public void init(SegmentGeneratorConfig config)
       throws Exception {
-    init(config, RecordReaderFactory.getRecordReader(config));
+    init(config, getRecordReader(config));
+  }
+
+  private RecordReader getRecordReader(SegmentGeneratorConfig segmentGeneratorConfig)
+      throws Exception {
+    File dataFile = new File(segmentGeneratorConfig.getInputFilePath());
+    Preconditions.checkState(dataFile.exists(), "Input file: " + dataFile.getAbsolutePath() + " does not exist");
+
+    Schema schema = segmentGeneratorConfig.getSchema();
+    FileFormat fileFormat = segmentGeneratorConfig.getFormat();
+    String recordReaderPath = segmentGeneratorConfig.getRecordReaderPath();
+
+    // Allow for instantiation general record readers from a record reader path passed into segment generator config
+    // If this is set, this will override the file format
+    if (recordReaderPath != null) {
+      if (fileFormat != FileFormat.OTHER) {
+        // NOTE: we currently have default file format set to AVRO inside segment generator config, do not want to break
+        // this behavior for clients.
+        LOGGER
+            .warn("Using class: {} to read segment, ignoring configured file format: {}", recordReaderPath, fileFormat);
+      }
+      return org.apache.pinot.spi.data.readers.RecordReaderFactory
+          .getRecordReaderByClass(recordReaderPath, dataFile, schema, segmentGeneratorConfig.getReaderConfig());
+    }
+
+    switch (fileFormat) {
```

Comment:
can we do PINOT here, and let getRecordReader add other file formats? Or, if I add a new file format, is this the only place to change?

Reply:
Good point, this will be much cleaner.